nomt/src/beatree/ops/update/branch_updater.rs

```diff
@@ -137,66 +183,147 @@ impl BranchUpdater {
-                    self.bulk_split_step(i);
+                let mut idx = 0;
+                while idx < self.ops.len() {
+                    let consumed = self.bulk_split_step(idx);
+                    idx += consumed;
                 }
+                idx
             }
             Some(ref mut bulk_splitter) if body_size_after >= BRANCH_BULK_SPLIT_TARGET => {
+                let mut recursive_on_split = false;
+
                 let accept_item = body_size_after <= BRANCH_NODE_BODY_SIZE || {
-                    if self.gauge.body_size() < BRANCH_MERGE_THRESHOLD {
-                        // rare case: body was artifically small due to long shared prefix.
-                        // start applying items without prefix compression. we assume items are less
-                        // than half the body size, so the next item should apply cleanly.
-                        self.gauge.stop_prefix_compression();
-                        true
-                    } else {
-                        false
+                    match self.ops[op_index] {
+                        BranchOp::Insert(..) | BranchOp::Update(..) => {
+                            if self.gauge.body_size() < BRANCH_MERGE_THRESHOLD {
+                                // rare case: body was artifically small due to long shared prefix.
+                                // start applying items without prefix compression. we assume items are less
+                                // than half the body size, so the next item should apply cleanly.
+                                self.gauge.stop_prefix_compression();
+                                true
+                            } else {
+                                false
+                            }
+                        }
+                        BranchOp::KeepChunk(from, to) => {
+                            let left_n_items = split_keep_chunk(
+                                self.base.as_ref().unwrap(),
+                                &self.gauge,
+                                &mut self.ops,
+                                op_index,
+                                BRANCH_NODE_BODY_SIZE,
+                                BRANCH_NODE_BODY_SIZE,
+                            );
+
+                            if left_n_items > 0 {
+                                // KeepChunk has been successfully split, the left chunk is now able to fit
+                                recursive_on_split = true;
+                                true
+                            } else {
+                                // Not even a single element was able to fit.
+                                if self.gauge.body_size() < BRANCH_MERGE_THRESHOLD {
+                                    // We can stop prefix compression and separate the first
+                                    // element of the keep_chunk into its own.
+                                    self.ops[op_index] = BranchOp::KeepChunk(from + 1, to);
```

Comment:
what about 0/1-sized chunks? Are they permitted? what are the invariants around `KeepChunk`?

Reply:
0-sized chunks should never be created, but the line of code you just commented on, as well as the previous comment, are two examples of possible cases of 0-sized chunk creations. I will correct this behavior in a follow-up, along with an explanation of the invariant `KeepChunk`